### PR TITLE
Add EventLog

### PR DESCRIPTION
**Pull request by tolemac**
This PR comes from some conversations on Discord.

At the first hand I needed to transform all tables from a database searching for 0x00 special chars on source, for this we add a `GetProcessor` method on `StreamTransformerBase` from `ITransformer` file.

The second thing was perform table mergin, on PostgreSQL, without any PK columns. If you don't have PK colums you don't know what to keep and what to update ... speaking on Discord accord to copy entire data and leave in user the possibility of delete data before sync. That is done only for PostgreSQL, I couldn't try on others.

At the last is something that came up talking about localize what tables have special chars. I needed continue on error feature, but I have 1800 tables in my database, the console output is not sufficient, then  I thought about have an event log where the pansynchro records what happens in the process, in order to save this information to disk at the end or ... sent by email ... what ever else.

### CODE REVIEW

**Comment by tolemac** [comment]
At the moment only errors are logged to event log. 
I'm working on replace all `Console.WriteLine("Message")` to `EventLog.AddInformationEvent("Message")`.

**Comment by tolemac** [comment]
After analize the outputs I have extracted this type of events:

* Sync data:
  * Start Sync All
  * Start Reading stream	
  * End Reading stream
  * Start writing Stream
  * End writing Stream
  * Start Merging Stream
  * End Merging Stream
  * Start Truncating all
  * End trucating all
  * Finnished Sync All

* Transformer: 
  * Processing stream
  * Stream found
  * Stream not found

Protocol (Binary decoder an encoder):
  * Bytes readed
  * Bytes written

I think the transformer output is excessive, I think on emit a event on start of transform an other when finished, and emit a event when a stream is found in transformer.

**Comment by tolemac** [comment]
I have added these events:
image

I think only protocol events are missing.

**Comment by tolemac** [comment]
No new events needed for protocol, only use the existings.

**Comment by tolemac** [comment]
Well the target of this PR is not to write to console directly, instead it now Pansynchro emits events throught a event system.
Event system store each event and hava a event list available to query or save it.

image
image

Each event fired with event system is "logged" to an `ILogger` (from Microsoft.Extensions.Logging).
By default the logger is initialized with a new instance of `ColorConsoleLogger` (custom ILogger in `Pansynchro.Core.Logs`), This custom logger write events to the console only if `EventLog.ConsoleOutput` flag is `true`.

image

You can set other ILogger if you want using `EventLog.SetDotNetLogger`. Or you can configure your DI to use ColocConsoleLogger as default logger.
image

At the last point you can enable  continue on error feature, when it is enable if some error raise when is synching the process go to the next stream instead stop the process. When process finish you can see errors in the event list o event system.

image

**Comment by tolemac** [comment]
I have squashed some commits and it's ready to review.

File: .editorconfig [hunk 1]
```diff
@@ -1,9 +1,164 @@
 ﻿[*.cs]
 
 # IDE0022: Use block body for methods
-csharp_style_expression_bodied_methods = false
+csharp_style_expression_bodied_methods = false:silent
```
Comment (masonwheeler) [inline .editorconfig]: Please don't include personal configuration files in the PR.

File: Pansynchro.Connectors.Postgres/MetadataHelper.cs [hunk 2]
```diff
@@ -160,11 +160,18 @@ internal static void MergeTable(NpgsqlConnection conn, string table, string name
...
-			if (nonPkColumns.Length == 0) {
-				MergeLinkingTable(conn, table, namespace_, columns, pkColumns);
-			} else {
-				MergeNormalTable(conn, table, namespace_, columns, pkColumns, nonPkColumns);
+			if (pkColumns.Length == 0)
+			{
+				CopyData(conn, table, namespace_, columns);
+			}
+			else
+			{
```
Comment (masonwheeler) [inline Pansynchro.Connectors.Postgres/MetadataHelper.cs]: Check the formatting conventions here.

File: Pansynchro.Connectors.Snowflake/SnowflakeUploader.cs [hunk 3]
```diff
@@ -80,12 +82,13 @@ private async Task UploadAsync(string filename, StreamDescription streamName)
                 var lFilename = Path.GetFileName(filename);
                 conn.Execute($"copy into {SnowflakeUploader.GetTargetName(streamName)} from '@{stageName}/{lFilename}' file_format = (type = avro) match_by_column_name = case_insensitive;");
                 await conn.CloseAsync();
-            } catch (Exception e) {
-                Console.WriteLine(e);
+            } catch (Exception e) {
```
Comment (masonwheeler) [inline Pansynchro.Connectors.Snowflake/SnowflakeUploader.cs]: For some strange reason, there's a lot of extra whitespace at the end of this line.

File: Pansynchro.Connectors.TextFile/CSV/CsvWriter.cs [hunk 4]
```diff
@@ -34,15 +36,21 @@ public async Task Sync(IAsyncEnumerable<DataStream> streams, DataDictionary dest
 			if (_sink == null) {
 				throw new DataException("Must call SetDataSink before calling Sync");
 			}
-			await foreach (var (name, settings, stream) in streams) {
+			EventLog.Instance.AddStartSyncEvent();
+            await foreach (var (name, settings, stream) in streams) {
```
Comment (masonwheeler) [inline Pansynchro.Connectors.TextFile/CSV/CsvWriter.cs]: Check the indentation here.

File: Pansynchro.Protocol/BinaryDecoder.cs [hunk 5]
```diff
@@ -222,7 +223,7 @@ private async IAsyncEnumerable<DataStream> ReadStreams(int length, DataDictionar
 				yield return new DataStream(streamName, StreamSettings.None, new StreamingReader(_reader, schema, dict, mode, () => locked = false));
 			}
 			CheckEqual(_reader.ReadByte(), (byte)Markers.End, "Expected end of stream not found.");
-			Console.WriteLine($"Bytes read: {_meter.TotalBytesRead.ToString("N0")}");
+            EventLog.Instance.AddInformationEvent($"Bytes read: {_meter.TotalBytesRead.ToString("N0")}");
```
Comment (masonwheeler) [inline Pansynchro.Protocol/BinaryDecoder.cs]: Check the indentation here.  This is admittedly not perfectly consistent, but I try to indent with tabs, one tab per indentation level.

File: Pansynchro.Protocol/BinaryEncoder.cs [hunk 6]
```diff
@@ -69,12 +71,15 @@ public BinaryEncoder(TcpListener server, DataDictionary dict, int compressionLev
...
+			EventLog.Instance.AddStartSyncEvent();
+
 			_dataDict ??= dest;
 			WriteVersion(_outputWriter);
 			WriteDataDictHash(_outputWriter, _dataDict);
 			using var bufferWriter = new BinaryWriter(_bufferStream, Encoding.UTF8);
 			await foreach (var (name, settings, reader) in streams) {
-				try {
+                EventLog.Instance.AddStartSyncStreamEvent(name);
+                try {
```
Comment (masonwheeler) [inline Pansynchro.Protocol/BinaryEncoder.cs]: More indentation fun in this file.

File: Pansynchro.core/EventsSystem/Events/StreamEvent.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Pansynchro.Core.EventsSystem.Events;
+public class StreamEvent : EventBase, IStreamEvent
+{
```
Comment (masonwheeler) [inline Pansynchro.core/EventsSystem/Events/StreamEvent.cs:9]: Should this have a custom `ToString()`?

File: Pansynchro.core/ITransformer.cs [hunk 8]
```diff
@@ -38,14 +39,19 @@ public virtual async IAsyncEnumerable<DataStream> StreamLast()
 			yield break;
 		}
 
-		public async IAsyncEnumerable<DataStream> Transform(IAsyncEnumerable<DataStream> input)
+        protected virtual Func<IDataReader, IEnumerable<object[]>>? GetProcessor(DataStream stream)
```
Comment (masonwheeler) [inline Pansynchro.core/ITransformer.cs]: Check the indentation on this file.

**Review by masonwheeler — CHANGES_REQUESTED** [review]
This looks like solid, well-written work.  Fix up a few minor cosmetic issues and I'll happily merge this.

File: Pansynchro.core/EventsSystem/Events/StreamEvent.cs (hunk 7 above)
Comment (tolemac) [reply Pansynchro.core/EventsSystem/Events/StreamEvent.cs:9]: The `ToString` method is implemented on `EventBase`, all cases are covered centrally there.

**Review by masonwheeler — APPROVED** [review]
Looks good!

**Comment by masonwheeler** [comment]
@tolemac Any idea what's going wrong with the [build failure?](https://github.com/Pansynchro-Technologies/Pansynchro/actions/runs/9903484201/job/27388227992?pr=1)

**Comment by tolemac** [comment]
@masonwheeler for some unknown reason git was ignoring `/Logs/ColorConsoleLogger` folder and some files were missing ...
I have uploaded the files in a new commit. I can understand what happends no changes on .gitignore ... I think is some issue with Visual Studio Git plugin.

Either way, I have upload the files now.

**Comment by masonwheeler** [comment]
Looks good!  I'll merge this, and set up a new NuGet release later today.

**Pull request merged**